### PR TITLE
test: add interfaces, mocks, and tests

### PR DESCRIPTION
**Pull request by bap2pecs**
## Summary
this is to complement #55 and is part of the issue #43 

more context see: https://github.com/babylonchain/babylon-finality-gadget/pull/45#discussion_r1666279613

> Suggest we formalise the mock procedure for unit tests, like what we did in finality provider

## Test Plan
```
make mock-gen
make lint
make test
```

### CODE REVIEW

File: sdk/client/interface.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package client
+
+import "github.com/babylonchain/babylon-finality-gadget/sdk/cwclient"
+
+type SdkClientInterface interface {
+	QueryIsBlockBabylonFinalized(queryParams *cwclient.L2Block) (bool, error)
+
+	QueryBlockRangeBabylonFinalized(queryBlocks []*cwclient.L2Block) (*uint64, error)
```
Comment (bap2pecs) [inline sdk/client/interface.go:8]: TODO: add docstring here

File: Makefile [hunk 2]
```diff
@@ -1,4 +1,17 @@
-.PHONY: lint test
+.PHONY: lint test mock-gen
+
+CUR_DIR := $(shell pwd)
+MOCKS_DIR=$(CUR_DIR)/testutil/mocks
+MOCKGEN_REPO=github.com/golang/mock/mockgen
+MOCKGEN_VERSION=latest
+MOCKGEN_CMD=mockgen
```
Comment (bap2pecs) [inline Makefile:7]: TODO: clean up

**Review by gitferry — COMMENTED** [review]
In general, I would suggest to put all the interfaces under `sdk/client/expected_clients.go`. This allows the actual user to define the interfaces that would be needed in one place. See how cosmos-sdk does it https://github.com/cosmos/cosmos-sdk/blob/1bd4351241783827d554d41edb3bbb383d587a97/x/slashing/types/expected_keepers.go#L15

File: sdk/bbnclient/interface.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package bbnclient
+
+type BBNClientInterface interface {
```
Comment (gitferry) [inline sdk/bbnclient/interface.go:3]: I don't think the Golang naming convention recommends to have `Interface` at the end for interfaces. I would recommend to name it something like `BabylonClientController`. This comment also applies to other interfaces named with `Interface`

Comment (bap2pecs) [reply sdk/bbnclient/interface.go]: good point. i like those new names

**Comment by bap2pecs** [comment]
resolved @gitferry 's comments to 1) move all interfaces to expected_clients.go 2) rename interfaces

File: sdk/client/interface.go [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package client
+
+import "github.com/babylonchain/babylon-finality-gadget/sdk/cwclient"
+
+type ISdkClient interface {
```
Comment (bap2pecs) [inline sdk/client/interface.go:5]: I want this to be in a separate file b/c I need to use the sdkclient mock in the FP repo

Comment (gitferry) [reply sdk/client/interface.go:5]: Good point

**Review by gitferry — APPROVED** [review]
Thanks for the changes! LGTM!

File: sdk/client/query_test.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+		{
+			name:           "everyone has 100 voting power, 3 of them votes, return true",
+			allFpPks:       []string{"pk1", "pk2", "pk3", "pk4"},
+			votedProviders: []string{"pk1", "pk2", "pk3"},
+			fpPowers:       map[string]uint64{"pk1": 100, "pk2": 100, "pk3": 100, "pk4": 100},
+			expectResult:   true,
+		},
+	}
+
+	fmt.Println("before test loop")
```
Comment (bap2pecs) [inline sdk/client/query_test.go:79]: TODO: remove those debug lines

**Pull request merged**